iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,327 @@
+use tokio::sync::Mutex;
+use tokio::time::Instant;
+
+use iroha_actor::{Actor, Context, Handler, MessageResponse, Addr, ContextHandler};
+use iroha_crypto::{Hash};
+use iroha_data_model::prelude::PeerId;
+use iroha_logger::{debug, error, info, trace, warn};
+
+use crate::message::{Message, MessageResult};
+use crate::network::{SendPeerMsg, ReceivedPeerMsg, Connect};
+use std::pin::Pin;
+use crate::Network;
+use rand::{Rng, RngCore};
+use crate::chacha::Chacha;
+use x25519_dalek::{PublicKey, StaticSecret};
+
+const MAX_MESSAGE_LENGTH: usize = 2 * 1024 * 1024;
+
+pub struct Peer {
+    id: PeerId,
+    read: Option<OwnedReadHalf>,
+    write: Option<OwnedWriteHalf>,
+    state: State,
+    secret_key: StaticSecret,
+    public_key: PublicKey,
+    cipher: Option<Chacha>,
+    network_addr: Addr<Network>,
+}
+
+impl Peer {
+    pub fn new(id: PeerId, stream: Option<TcpStream>, state: State, addr: Addr<Network>) -> Self {
+        // P2P encryption primitives
+        let mut thread_rng = rand::thread_rng();
+        let secret_key = StaticSecret::new(&mut thread_rng);
+        let public_key = PublicKey::from(&secret_key);
+        // If we are connected we take apart stream for two halves.
+        // If we are not connected we save Nones and wait for message to start connecting.
+        match stream {
+            None => {
+                Self { id, read: None, write: None, state, secret_key, public_key, cipher: None, network_addr: addr }
+            }
+            Some(stream) => {
+                let (read, write) = stream.into_split();
+                Self { id, read: Some(read), write: Some(write), state, secret_key, public_key, cipher: None, network_addr: addr }
+            }
+        }
+
+    }
+
+    fn stream(&mut self) -> impl Stream<Item = MessageResult> + Send + 'static {
+        let mut read: OwnedReadHalf = self.read.take().unwrap();
+        stream! {
+            loop {
+                if let Err(e) = readable(&read).await {
+                    yield MessageResult::new_error(e);
+                    break;
+                }
+                let result = match read_message(&mut read).await {
+                    Ok(message) => MessageResult::new_message(message),
+                    Err(e) => MessageResult::new_error(e)
+                };
+
+                yield result;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Actor for Peer {
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        //self.addr = Some(ctx.addr());
+        // Subscribe reading stream
+        ctx.notify_with(self.stream());
+    }
+}
+
+#[async_trait::async_trait]
+impl Handler<MessageResult> for Peer {
+    type Result = ();
+
+    async fn handle(&mut self, msg: MessageResult) {
+        match msg {
+            MessageResult::Ok { message } => {
+                // We got handshake or the real message from other peer
+                match message {
+                    Message::ClientHello { key} => {
```

Comment:
May be you could move each of the match branch to some method of type? That would be easier to read